# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources (README.md file)
- Happy learning!

In [41]:
# import numpy and pandas
import pandas as pd
import numpy as np
from scipy.stats import trim_mean, mode, skew, gaussian_kde, pearsonr, spearmanr, beta
from statsmodels.stats.weightstats import ztest as ztest
from statsmodels.stats.proportion import proportions_ztest

from scipy.stats import ttest_ind, norm, t
from scipy.stats import ttest_1samp
from scipy.stats import f_oneway
from scipy.stats import sem


# Challenge 1 - Exploring the Data

In this challenge, we will examine all salaries of employees of the City of Chicago. We will start by loading the dataset and examining its contents

In [42]:
# Run this code:
salaries = pd.read_csv('../data/Current_Employee_Names__Salaries__and_Position_Titles.csv')


Examine the `salaries` dataset using the `head` function below.

In [43]:
salaries.head()

,Name,Job Titles,Department,Full or Part-Time,Salary or Hourly,Typical Hours,Annual Salary,Hourly Rate
0,"AARON, JEFFERY M",SERGEANT,POLICE,F,Salary,NaN,101442.0,NaN
1,"AARON, KARINA",POLICE OFFICER (ASSIGNED AS DETECTIVE),POLICE,F,Salary,NaN,94122.0,NaN
2,"AARON, KIMBERLEI R",CHIEF CONTRACT EXPEDITER,GENERAL SERVICES,F,Salary,NaN,101592.0,NaN
3,"ABAD JR, VICENTE M",CIVIL ENGINEER IV,WATER MGMNT,F,Salary,NaN,110064.0,NaN
4,"ABASCAL, REECE E",TRAFFIC CONTROL AIDE-HOURLY,OEMC,P,Hourly,20.0,NaN,19.86


# Challenge 2 - Hypothesis Tests

In this section of the lab, we will test whether the hourly wage of all hourly workers is significantly different from $30/hr. Import the correct one sample test function from scipy and perform the hypothesis test for a 95% two sided confidence interval.

$ H_0 : \mu = 30 $  
$ H_1 : \mu \neq 30 $

In [44]:
salaries_hourlyrate = salaries["Hourly Rate"].dropna()
ttest_1samp(salaries_hourlyrate, popmean=30, alternative="two-sided")

TtestResult(statistic=np.float64(20.6198057854942), pvalue=np.float64(4.323024048623004e-92), df=np.int64(8021))

**__Relative to standard error__** the observed hourly rate deviates significantly from $30 per hour, with a p value of essentially zero. which is sufficient to reject the null hypothesis $H_0$.  
  
We can also confirm this by taking the mean of non null Hourly Rate values:

In [45]:
salaries_hourlyrate.mean()

np.float64(32.78855771628024)

The mean being 32.79 (2dp) does not seem like a large deiviation but remeber the deivation is large relative to standard error, due to the number of values in the distribution.

# Challenge 3 - Constructing Confidence Intervals

While testing our hypothesis is a great way to gather empirical evidence for accepting or rejecting the hypothesis, another way to gather evidence is by creating a confidence interval. A confidence interval gives us information about the true mean of the population. So for a 95% confidence interval, we are 95% sure that the mean of the population is within the confidence interval. 
).

To read more about confidence intervals, click [here](https://en.wikipedia.org/wiki/Confidence_interval).


In the cell below, we will construct a 95% confidence interval for the mean hourly wage of all hourly workers. 

The confidence interval is computed in SciPy using the `t.interval` function. You can read more about this function [here](https://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.stats.t.html).

To compute the confidence interval of the hourly wage, use the 0.95 for the confidence level, number of rows - 1 for degrees of freedom, the mean of the sample for the location parameter and the standard error for the scale. The standard error can be computed using [this](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.sem.html) function in SciPy.

In [46]:
t.interval(
    confidence =0.95,                                       ## 95% confidence interval
    df = (len(salaries_hourlyrate)-1),                      ## Degrees of freedom = n-1
    loc=salaries_hourlyrate.mean(),                         ## loc (centre of distribution) = mean
    scale = sem(salaries_hourlyrate)                        ## sem provides a estimate of Standard error from one sample, based on standard deviation and the assumptions that the sample is large enough for CLT to apply, representative of the population, and a random selection of the population
    )

(np.float64(32.52345834488425), np.float64(33.05365708767623))

Standard error of the mean is used as the scale not standard deviation as it scales with the size of our sample.Standard deviation would not so would not reflect the increased confidence form a larger sample.

# Challenge 4 - Hypothesis Tests of Proportions

Another type of one sample test is a hypothesis test of proportions. In this test, we examine whether the proportion of a group in our sample is significantly different than a fraction. 

You can read more about one sample proportion tests [here](http://sphweb.bumc.bu.edu/otlt/MPH-Modules/BS/SAS/SAS6-CategoricalData/SAS6-CategoricalData2.html).

In the cell below, use the `proportions_ztest` function from `statsmodels` to perform a hypothesis test that will determine whether the number of hourly workers in the City of Chicago is significantly different from 25% at the 95% confidence level.

$H_0 :$ Proportion of hourly paid Chicago workforce $ = 25 \% $  
$H_1 :$ Proportion of hourly paid Chicago workforce $ \neq 25 \% $


In [53]:
number_hourly_workforce = salaries_hourlyrate.shape[0]
number_total_workforce = salaries.shape[0]
percentage_of_workforce_hourly = number_hourly_workforce / number_total_workforce
percentage_of_workforce_hourly

0.24175029382515142

Two-sided hypothesis test

In [54]:
stat , pval = proportions_ztest(
    count = number_hourly_workforce,
    nobs = number_total_workforce,
    value = 0.25,
    alternative = "two-sided"
    )

In [55]:
print(f"""
zstat = {stat}
pval = {pval}
""")


zstat = -3.5099964213703005
pval = 0.0004481127249057967



Our observed value is aproximately 3.5 estimated standard errors less than the expected value from $H_0$ with a pvalue of 0.0004 (4dp) to acompany it.  
  
This is sufficient to reject our null hypothesis that 25% of the workforce is paid hourly.